File: python/otel_helper/metrics.py
"""Metrics setup — equivalent to MetricsSetup.cs."""

import logging

from opentelemetry import metrics
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.metrics._internal.exemplar import TraceBasedExemplarFilter
from opentelemetry.sdk.metrics.view import View, DropAggregation
from opentelemetry.sdk.resources import Resource

from otel_helper.config import TelemetryOptions

_logger = logging.getLogger(__name__)


def _build_drop_views(patterns: list[str]) -> list[View]:
    """Build View objects that drop metrics matching wildcard patterns."""
    views = []
    for pattern in patterns:
        views.append(View(instrument_name=pattern, aggregation=DropAggregation()))
    return views


def metrics_app():
    """ASGI app serving the Prometheus scrape endpoint.

    Mount it on the application's own server — the correct mode under
    multi-worker servers (gunicorn/uvicorn workers), where the standalone
    listener cannot bind the same port in every process:

        app.mount("/metrics", metrics_app())

    Combine with OTEL_HELPER_METRICS_PORT=0 to suppress the standalone listener.
    """
    from prometheus_client import make_asgi_app

    return make_asgi_app()


def configure_metrics(resource: Resource, options: TelemetryOptions) -> MeterProvider:
    """Configure and set the global MeterProvider with trace-based exemplars.

    The active exporters come from options.resolved_metric_exporters():
    "otlp" (push) and/or "prometheus" (/metrics scrape) on the same provider.
    """
    exporters = options.resolved_metric_exporters()
    readers = []

    if "otlp" in exporters:
        from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter

        exporter = OTLPMetricExporter(
            endpoint=options.otel_endpoint,
            insecure=options.resolve_insecure(),
            timeout=options.export_timeout_ms / 1000,
        )
        readers.append(PeriodicExportingMetricReader(
            exporter, export_interval_millis=options.export_interval_ms,
        ))

    if "prometheus" in exporters:
        from opentelemetry.exporter.prometheus import PrometheusMetricReader

        readers.append(PrometheusMetricReader())
        if options.prometheus_metrics_port > 0:
            from prometheus_client import start_http_server

            try:
                start_http_server(port=options.prometheus_metrics_port)
            except OSError as e:
                raise RuntimeError(
                    f"[OtelHelper] Could not bind the /metrics listener on port "
                    f"{options.prometheus_metrics_port}: {e}. If another process (or worker) owns "
                    "the port, mount otel_helper.metrics_app() on your app and set "
                    "OTEL_HELPER_METRICS_PORT=0 to disable the standalone listener."
                ) from e
            _logger.info(
                "OTel Prometheus exporter: metrics available at http://0.0.0.0:%d/metrics",
                options.prometheus_metrics_port,
            )
        else:
            _logger.info(
                "OTel Prometheus exporter: standalone listener disabled (port=0); "
                "mount otel_helper.metrics_app() to expose /metrics",
            )

    views = _build_drop_views(options.disabled_metrics) if options.disabled_metrics else []

    kwargs = dict(
        resource=resource,
        metric_readers=readers,
        exemplar_filter=TraceBasedExemplarFilter(),
    )
    if views:
        kwargs["views"] = views

    provider = MeterProvider(**kwargs)
    metrics.set_meter_provider(provider)
    return provider

File: go/metrics.go
package otelhelper

import (
	"context"
	"crypto/tls"
	"errors"
	"fmt"
	"net"
	"net/http"
	"os"
	"path"
	"sync"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"go.opentelemetry.io/contrib/instrumentation/runtime"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	promexporter "go.opentelemetry.io/otel/exporters/prometheus"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	"google.golang.org/grpc/credentials"
)

// promRegistry is the dedicated registry backing the Prometheus reader.
// Deliberately NOT the client_golang global registry, so application metrics
// registered there don't leak into the helper's /metrics and vice versa.
var (
	promRegistryMu sync.Mutex
	promRegistry   *prometheus.Registry
)

// MetricsHandler returns an http.Handler serving the Prometheus scrape
// endpoint. Mount it on the application's own mux:
//
//	mux.Handle("/metrics", otelhelper.MetricsHandler())
//
// Combine with WithoutMetricsListener() (or OTEL_HELPER_METRICS_PORT=0) to
// suppress the standalone listener. Before Setup configures the Prometheus
// reader, the handler answers 503.
func MetricsHandler() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		promRegistryMu.Lock()
		reg := promRegistry
		promRegistryMu.Unlock()
		if reg == nil {
			http.Error(w, "otelhelper: prometheus metrics reader not configured", http.StatusServiceUnavailable)
			return
		}
		promhttp.HandlerFor(reg, promhttp.HandlerOpts{}).ServeHTTP(w, r)
	})
}

func setPromRegistry(reg *prometheus.Registry) {
	promRegistryMu.Lock()
	defer promRegistryMu.Unlock()
	promRegistry = reg
}

// configureMetrics builds the MeterProvider with one reader per resolved
// exporter ("otlp" push and/or "prometheus" scrape — design D2). The returned
// shutdown (possibly nil) stops the standalone /metrics listener.
func configureMetrics(ctx context.Context, res *resource.Resource, opts *Options) (*sdkmetric.MeterProvider, func(context.Context) error, error) {
	// TODO: Replace with sdkmetric.WithExemplarFilter(sdkmetric.TraceBasedExemplarFilter)
	// when available in a future SDK version. The programmatic API is not exported in v1.31.0.
	// This must be called before spawning goroutines (Setup is called early at startup).
	if os.Getenv("OTEL_METRICS_EXEMPLAR_FILTER") == "" {
		os.Setenv("OTEL_METRICS_EXEMPLAR_FILTER", "trace_based")
	}

	mpOpts := []sdkmetric.Option{sdkmetric.WithResource(res)}
	var srvShutdown func(context.Context) error

	if opts.hasMetricExporter("otlp") {
		metricOpts := []otlpmetricgrpc.Option{
			otlpmetricgrpc.WithEndpoint(opts.OtelEndpoint),
			otlpmetricgrpc.WithCompressor("gzip"),
		}
		if opts.Insecure {
			metricOpts = append(metricOpts, otlpmetricgrpc.WithInsecure())
		} else {
			metricOpts = append(metricOpts, otlpmetricgrpc.WithTLSCredentials(credentials.NewTLS(&tls.Config{})))
		}
		exporter, err := otlpmetricgrpc.New(ctx, metricOpts...)
		if err != nil {
			return nil, nil, fmt.Errorf("metric exporter: %w", err)
		}
		interval := opts.ExportIntervalMs
		if interval <= 0 {
			interval = defaultExportIntervalMs
		}
		mpOpts = append(mpOpts, sdkmetric.WithReader(sdkmetric.NewPeriodicReader(exporter,
			sdkmetric.WithInterval(time.Duration(interval)*time.Millisecond),
		)))
	}

	if opts.hasMetricExporter("prometheus") {
		registry := prometheus.NewRegistry()
		exporter, err := promexporter.New(promexporter.WithRegisterer(registry))
		if err != nil {
			return nil, nil, fmt.Errorf("prometheus exporter: %w", err)
		}
		setPromRegistry(registry)
		mpOpts = append(mpOpts, sdkmetric.WithReader(exporter))

		if opts.PrometheusMetricsPort > 0 {
			// Synchronous Listen: a busy port fails Setup immediately instead of
			// dying silently in a goroutine (the P5 "silent telemetry loss" bug).
			ln, err := net.Listen("tcp", fmt.Sprintf(":%d", opts.PrometheusMetricsPort))
			if err != nil {
				return nil, nil, fmt.Errorf(
					"prometheus /metrics listener on port %d: %w (if another process owns the port, mount otelhelper.MetricsHandler() and use WithoutMetricsListener())",
					opts.PrometheusMetricsPort, err)
			}
			mux := http.NewServeMux()
			mux.Handle("/metrics", promhttp.HandlerFor(registry, promhttp.HandlerOpts{}))
			srv := &http.Server{
				Handler:           mux,
				ReadHeaderTimeout: 5 * time.Second,
			}
			go func() {
				if err := srv.Serve(ln); err != nil && !errors.Is(err, http.ErrServerClosed) {
					otel.Handle(fmt.Errorf("otelhelper: prometheus /metrics server: %w", err))
				}
			}()
			srvShutdown = srv.Shutdown
		}
	}

	if len(opts.DisabledMetrics) > 0 {
		mpOpts = append(mpOpts, sdkmetric.WithView(disabledMetricsView(opts.DisabledMetrics)))
	}

	mp := sdkmetric.NewMeterProvider(mpOpts...)
	otel.SetMeterProvider(mp)

	// Start runtime metrics (goroutines, GC, memory). Non-fatal if it fails.
	if err := runtime.Start(runtime.WithMeterProvider(mp)); err != nil {
		otel.Handle(err)
	}

	return mp, srvShutdown, nil
}

// disabledMetricsView returns a View that drops metrics matching any of the given glob patterns.
func disabledMetricsView(patterns []string) sdkmetric.View {
	return func(inst sdkmetric.Instrument) (sdkmetric.Stream, bool) {
		for _, pattern := range patterns {
			if matched, _ := path.Match(pattern, inst.Name); matched {
				return sdkmetric.Stream{Aggregation: sdkmetric.AggregationDrop{}}, true
			}
		}
		return sdkmetric.Stream{}, false
	}
}
